20251028-BRS-workflow-versioning.md

```diff
@@ -0,0 +1,696 @@
+
+### Version isn't specified on orchestration request
+If there is no version specified on the event, the SDK should map the request based on the named base type to the
+latest version available in the application and run the workflow there. Every time a patch is evaluated, the "new code" 
+route should be taken, and the patch identifier persisted in a list of encountered patches.
+
+At each boundary invocation except workflow orchestration completion, the SDK should return the workflow status with
+the (potentially updated) version information defined below containing the name of the workflow type executed and the list 
+of named patches followed.
+
+### Version is specified on orchestration request
+If the version is supplied, the workflow type should be parsed out and the SDK should ignore "latest" evaluations and use
+that type to run the workflow execution. The list of patches should be extracted from the version prototype. At each
+evaluation of the patch, the path should be taken that corresponds to whether the patch exists in the list:
+
+- If the patch exists in the version list, the "new code" route should be taken
+- If the patch does not exist in the version list, the "old code" route should be taken
+
+#### Version Specification
+The version object needs to reflect the name (not the canonical name) of the workflow that was that actually invoked
+in the previous execution so the SDKs can override the routing and point directly to this type. It also needs to include
+the list of patch identifiers that evaluated as true when this was run so they might replay in the same manner.
+
+We implement this in the following manner instead of using a fixed versioning prototype because the list of patch names 
+should only be populated with the new patch identifiers encountered each workflow execution and should not be a running 
+list as this would be unnecessarily redundant and a waste of state resources. This approach also means that versioning
+is strictly stored in the workflow history state and doesn't require any change to how the runtime starts orchestration
+requests (via `OrchestratorRequest` messages).
+
+The following modifications should be made on the `OrchestratorCompletedEvent` when the SDK has finished workflow execution
+so it can reflect the name of the workflow type that was executed along with the list of patch identifiers that were 
+newly evaluated as true during the execution. Both properties are marked as optional for compatibility with older
+events, but in practice, the SDK should _always_ populate at least the `workflowTypeName` and the `patchNames` as 
+applicable for a versioned workflow going forward. 
+
+```protos
+message OrchestratorCompletedEvent {
+    // The name of the specific workflow type executed
+    optional workflowTypeName = 1;
+    // The list of patches that newly evaluated as true during the orchestrator execution
+    optional repeated string patchNames = 2;
```

Comment:
`repeated` fields can't be `optional`, if missing it'll just return an empty array, so we won't know if it's empty because it's really empty or because it's missing
If we want to be explicit with the presence (I'm not sure we strictly need to, but it's probably a good idea), we will need to put the patches in a separate message just for the repeated string.
Like this:
```protobuf
message PatchInformation {
  repeated string patchNames = 1;
}
message OrchestratorCompletedEvent {
  //...
  optional PatchInformation patchInformation = 2;
}
```